### PR TITLE
Batons are now Upgradeable - 1 Hit Stun - Minshield Firing Pin

### PR DESCRIPTION
**Pull request by JamieD1**
Batons are now upgradeable, you can now upgrade batons with a module which is printable via research to make them 1 hit Stun.(Advanced Weapon Research)
They now come with firing pins also which can be swapped out to lock them to sec only.

This is my compromise after weeks of discussions with council, players, staff and developers. If you can think of a better one feel free to give me it, but I am honestly fed up with having to think about batons. 

Just to add this is an OR upgrade, they can be 1 hit or Mind shielded apparently now.

Closes: #14061
Merges: #14322
:cl:  
rscadd: Batons are now upgradable to be 1 hit stun and can also be equipped with mindshield firing pins
/:cl:

### CODE REVIEW

File: code/game/objects/items/stunbaton.dm [hunk 1]
```diff
@@ -124,6 +152,11 @@
 		return ..()
 
 /obj/item/melee/baton/attack_self(mob/user)
+	if(!pin)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: remove

File: code/game/objects/items/stunbaton.dm [hunk 2]
```diff
@@ -143,6 +176,12 @@
 	add_fingerprint(user)
 
 /obj/item/melee/baton/attack(mob/M, mob/living/carbon/human/user)
+	if(!pin)
+		to_chat("There is no firing pin installed!")
+		return FALSE
+	if(!handle_pins(user))
+		to_chat(user, span_warning("You are not authorised to use [src]."))
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: remove

**Comment by ToasterBiome** [comment]
where charge batong

File: code/game/objects/items/stunbaton.dm [hunk 3]
```diff
@@ -143,6 +176,12 @@
 	add_fingerprint(user)
 
 /obj/item/melee/baton/attack(mob/M, mob/living/carbon/human/user)
+	if(!pin)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: remove

File: code/game/objects/items/stunbaton.dm [hunk 4]
```diff
@@ -124,6 +152,11 @@
 		return ..()
 
 /obj/item/melee/baton/attack_self(mob/user)
+	if(!pin)
+		to_chat("There is no firing pin installed!")
+	if(!handle_pins(user))
+		to_chat(user, span_warning("You are not authorised to use [src]."))
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: remove

File: code/game/objects/items/stunbaton.dm [hunk 5]
```diff
@@ -110,7 +125,20 @@
...
+	else if(istype(W, /obj/item/firing_pin))
+		if(pin)
+			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar"))
+	else if(W.tool_behaviour == TOOL_CROWBAR)
+		if(pin)
+			pin.forceMove(get_turf(src))
+			pin = null
+			status = FALSE
+			to_chat(user, span_notice("You remove the firing pin from [src]."))
+		if(upgrade)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm:184]: remove this, make the upgrade just change the stats once, and then delete the upgrade.
nobody is going to un-upgrade their baton

**Review by Bibby0110 — CHANGES_REQUESTED** [review]

**Comment by ynot01** [comment]
one click win is a cancer upon this codebase

**Comment by Hopekz** [comment]
Never even thought about sec firing pins for batons. Interesting

**Comment by ShadowDeath6** [comment]
Is there a way to just have it stun like it used to without shitting out stamina damage?

**Comment by JamieD1** [comment]
> Is there a way to just have it stun like it used to without shitting out stamina damage?

Can do if that is a valid concern

**Comment by ShadowDeath6** [comment]
I haven't played in a bit so I'm not sure if stamina damage's movespeed penalty has been changed, but I'd prefer to not move at a crawl if sec doesn't manage to cuff me by the time the stun ends.

File: code/game/objects/items/stunbaton.dm (hunk 5 above)
Comment (JamieD1) [reply code/game/objects/items/stunbaton.dm:184]: They might want to due to the extra power draw.

Comment (Bibby0110) [reply code/game/objects/items/stunbaton.dm:184]: thats the tradeoff

Comment (JamieD1) [reply code/game/objects/items/stunbaton.dm:184]: As batons are not easily obtainable I cant make them disposable like that.

**Comment by stwordcitizen** [comment]
overpowered

File: code/game/objects/items/stunbaton.dm [hunk 6]
```diff
@@ -44,13 +46,26 @@
 
 /obj/item/melee/baton/Initialize()
 	. = ..()
+	status = FALSE
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: change the declaration

File: code/game/objects/items/stunbaton.dm [hunk 7]
```diff
@@ -44,13 +46,26 @@
...
 	if(preload_cell_type)
 		if(!ispath(preload_cell_type,/obj/item/stock_parts/cell))
 			log_mapping("[src] at [AREACOORD(src)] had an invalid preload_cell_type: [preload_cell_type].")
 		else
 			cell = new preload_cell_type(src)
+	if(pin)
+		pin = new pin(src)
 	update_icon()
 
+/obj/item/melee/baton/Destroy()
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm:63]: delete the upgrade

File: code/game/objects/items/stunbaton.dm [hunk 8]
```diff
@@ -44,13 +46,26 @@
...
+	. = ..()
+	if(isobj(pin)) //Can still be the initial path, then we skip
+		QDEL_NULL(pin)
+
+/obj/item/melee/baton/handle_atom_del(atom/A)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm:70]: ditto

File: code/game/objects/items/stunbaton.dm [hunk 9]
```diff
@@ -110,7 +125,37 @@
 			cell = W
 			to_chat(user, span_notice("You install a cell in [src]."))
 			update_icon()
-
+	else if(istype(W, /obj/item/firing_pin))
+		if(upgrade)
+			if(W.type != /obj/item/firing_pin)
+				to_chat("You are unable to add a non default firing pin whilst it has an upgrade, remove it first with a crowbar")
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
				to_chat("You are unable to add a non default firing pin whilst [src] has an upgrade. Remove the upgrade first with a crowbar.")
```

File: code/game/objects/items/stunbaton.dm [hunk 10]
```diff
@@ -110,7 +125,37 @@
...
+				return
+		if(pin)
+			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar"))
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar."))
```

File: code/game/objects/items/stunbaton.dm [hunk 11]
```diff
@@ -110,7 +125,37 @@
...
+		else
+			W.forceMove(src)
+			pin = W
+	else if(istype(W, upgrade))
+		if(pin)
+			if(pin.type != /obj/item/firing_pin)
+				to_chat("You are unable to upgrade the baton whilst it has a non default pin")
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
```

File: code/game/objects/items/stunbaton.dm [hunk 12]
```diff
@@ -110,7 +125,37 @@
...
+				return
+		if(upgrade)
+			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
+		else
+			W.forceMove(src)
+			upgrade = W	
+	else if(W.tool_behaviour == TOOL_CROWBAR)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm:176]: call update icon somewhere here

File: code/game/objects/items/stunbaton.dm [hunk 13]
```diff
@@ -124,6 +169,10 @@
 		return ..()
 
 /obj/item/melee/baton/attack_self(mob/user)
+	if(!pin)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]: remove

File: code/game/objects/items/stunbaton.dm [hunk 14]
```diff
@@ -283,3 +340,20 @@
...
+			return TRUE
+		else
+			pin.auth_fail(user)
+			return FALSE
+	else
+		to_chat(user, span_warning("[src]'s trigger is locked. This weapon doesn't have a firing pin installed!"))
+	return FALSE
+
+/obj/item/batonupgrade
+	name = "Baton power upgrade"
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
	name = "baton power upgrade"
```

File: code/game/objects/items/stunbaton.dm [hunk 15]
```diff
@@ -283,3 +340,20 @@
...
+	desc = "A new power management circuit which enables double the power drain to instant stun."
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
	desc = "A new power management circuit which enables stun batons to instantly stun, at the cost of double power usage."
```

File: code/modules/research/designs/weapon_designs.dm [hunk 16]
```diff
@@ -104,6 +104,16 @@
 	category = list("Firing Pins")
 	departmental_flags = DEPARTMENTAL_FLAG_SECURITY
 
+/datum/design/batonupgrade
+	name = "Baton Power Upgrade"
+	desc = "A new power management circuit which enables double the power drain to instant stun."
```
Comment (Bibby0110) [inline code/modules/research/designs/weapon_designs.dm]:
```suggestion
	desc = "A new power management circuit which enables stun batons to instantly stun, at the cost of double power usage."
```

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
make it not apply to the crafted stunprod

**Comment by Hopekz** [comment]
> overpowered

How so?

**Comment by SomeguyManperson** [comment]
no

**Comment by maxion12345** [comment]
why the fuck would I use a mindshield baton, i just carry my trusty booby-trapped Decoy Baton to blow up anyone who tries to attack me lmao

**Comment by Hopekz** [comment]
> why the fuck would I use a mindshield baton, i just carry my trusty booby-trapped Decoy Baton to blow up anyone who tries to attack me lmao

Nice very specific anecdotal example that doesn't apply to most players

**Comment by Marmio64** [comment]
what about instead of upgrading a normal sized item into a one click win button you can instead make a new baton item from the fabricator (edit: or just make it an upgrade which does the same to an existing baton, just have to make sure the upgrade itself is also bulky) which is bulky (still fits in sec belts though) but has the instant stun capability. Could possibly alleviate frustration when antags get them and hide them in bags so you have no indication of whether or not they can instantly end you.

**Comment by ghost** [comment]
Lets do some math:

For example, changeling armblade(because changelings are recognised as strong antagonists so yeah:
How many attacks to win - 4
Seeing it makes all people validhunt and try to kill you - Yes
Needs more attacks to win if target is armored - Yes
Can be upgraded to allow it win in one hit - No

Funny stunbatton:
How many attacks to win - 2
Seeing it makes all people validhunt and try to kill you - Mostly no
Needs more attack to win if the target is armored - No, you litteraly can win a deathsquad officer by hitting them two times with a baton, and even his OP armor can't do anything about it
Can be upgraded to allow it win in one hit - with this PR - yes

Balance

When returning tasers back to the game?

**Comment by Marmio64** [comment]
> Lets do some math:
> 
> For example, changeling armblade(because changelings are recognised as strong antagonists so yeah: How many attacks to win - 4 Seeing it makes all people validhunt and try to kill you - Yes Needs more attacks to win if target is armored - Yes Can be upgraded to allow it win in one hit - No
> 
> Funny stunbatton: How many attacks to win - 2 Seeing it makes all people validhunt and try to kill you - Mostly no Needs more attack to win if the target is armored - No, you litteraly can win a deathsquad officer by hitting them two times with a baton, and even his OP armor can't do anything about it Can be upgraded to allow it win in one hit - with this PR - yes
> 
> Balance
> 
> When returning tasers back to the game?

partly wrong and unfair assessment/analogy here.
1: Changelings are recognized as a strong antagonist, not their armblade ability, which is actually sometimes considered a noobtrap BECAUSE of it outing you as a ling.
2: stunbaton stam damage is reduced by energy armor. Though energy armor is often really low, deathsquad armor does halve it, which means it does take 3 stunbaton hits to down them.
3: stunbatons have a 1.2 second cooldown between stuns. Melee cooldown is 0.8 seconds. Theoretically, in the time it takes for you to baton someone twice they can melee you three times.
4: stunbatons need to be earned after killing an officer, unlike say an armblade, which you can right off the bat get as a ling. Sec officers also dont have the element of surprise most of the time, so they should have stronger gear to compensate. 
edit: a fifth but rather small point. Batons need charge and have to be turned on before use unless you want to waste that charge.

not necessarily saying i agree with the PR or that batons are balanced but just pointing out some things.

**Comment by ghost** [comment]
> > Lets do some math:
> > 
> > For example, changeling armblade(because changelings are recognised as strong antagonists so yeah: How many attacks to win - 4 Seeing it makes all people validhunt and try to kill you - Yes Needs more attacks to win if target is armored - Yes Can be upgraded to allow it win in one hit - No
> > 
> > Funny stunbatton: How many attacks to win - 2 Seeing it makes all people validhunt and try to kill you - Mostly no Needs more attack to win if the target is armored - No, you litteraly can win a deathsquad officer by hitting them two times with a baton, and even his OP armor can't do anything about it Can be upgraded to allow it win in one hit - with this PR - yes
> > 
> > Balance
> > 
> > When returning tasers back to the game?
> 
> partly wrong and unfair assessment/analogy here.
> 1: Changelings are recognized as a strong antagonist, not their armblade ability, which is actually sometimes considered a noobtrap BECAUSE of it outing you as a ling.
> 2: stunbaton stam damage is reduced by energy armor. Though energy armor is often really low, deathsquad armor does halve it, which means it does take 3 stunbaton hits to down them.
> 3: stunbatons have a 1.2 second cooldown between stuns. Melee cooldown is 0.8 seconds. Theoretically, in the time it takes for you to baton someone twice they can melee you three times.
> 4: stunbatons need to be earned after killing an officer, unlike say an armblade, which you can right off the bat get as a ling. Sec officers also dont have the element of surprise most of the time, so they should have stronger gear to compensate. 
> edit: a fifth but rather small point. Batons need charge and have to be turned on before use unless you want to waste that charge.
> 
> not necessarily saying i agree with the PR or that batons are balanced but just pointing out some things.


This PR still sucks, it kills almost all stealth, non murderbone antagonists. Have fun waiting many rounds to roll an antagonist, and then getting your round ruined by a one lucky click of a stunbatton.

**Comment by SomeguyManperson** [comment]
one hit batons are always stupid and cannot be less stupid. Antags will  use them instead of antag specific gear because they are more reliable and prevent people from finding out exactly what their other antag gear is, sec will always use them because it's brainless to just run at people and hit them once with no counterplay outside "dont get hit lmao", and it gives anyone who has one this stupid "i can take 5 guys at once" mentality where they WILL run into stupid fights because there is a small chance they can click on people faster than they get clicked on (especially if nobody else present has an instant win button)

Also doubled charge cost isn't anywhere near enough of a penalty  to make it not worth taking every time it's available over literally anything else

**Comment by Hopekz** [comment]
> just have to make sure the upgrade itself is also bulky) which is bulky (still fits in sec belts though) 

I don't get this reasoning tbh.

If an item is "bulky" but has a convenient and common way to store it that everyone using it has (a security belt) then it is effectively not bulky because nothing has changed.

Unless you're trying to make it so antags can't put the stun batons in their bag unless they take a security belt? (not sure if belts can be hidden in backpacks or if it was finally fixed)

**Comment by ghost** [comment]
I suggest to make batons instanly dust people with antag datums, assistants and clowns

**Comment by maxion12345** [comment]
>one hit batons return
>security gets credits from harmbaton

Oh yeah, it's all coming together...

**Comment by Marmio64** [comment]
> > just have to make sure the upgrade itself is also bulky) which is bulky (still fits in sec belts though)
> 
> I don't get this reasoning tbh.
> 
> If an item is "bulky" but has a convenient and common way to store it that everyone using it has (a security belt) then it is effectively not bulky because nothing has changed.
> 
> Unless you're trying to make it so antags can't put the stun batons in their bag unless they take a security belt? (not sure if belts can be hidden in backpacks or if it was finally fixed)

i stated my reasoning in the same comment. " Could possibly alleviate frustration when antags get them and hide them in bags so you have no indication of whether or not they can instantly end you."
Though I did forget that sec belts are the one belts that aren't bulky so it is null. Though I suppose you could try to solve it by making it so the baton has to be equipped in the armor slot so it's guaranteed people will know if someone can one-shot them or not.

**Comment by AnCopper** [comment]
Why... why don't you people just use stamina batons?

**Comment by JamieD1** [comment]
> Why... why don't you people just use stamina batons?

Hi, you obviously are not from around here.

**Comment by Hopekz** [comment]
> > Why... why don't you people just use stamina batons?
> 
> Hi, you obviously are not from around here.

He does have a point though.
Like what's stopping us from just not changing anything.

What are the actual problems that need fixing that had you "discuss for weeks" about?

**Comment by ghost** [comment]
Not changing anything is better then removing all the fun from playing as an antagonist by making funny batons(that are already the strongest weapon in the game after tasers and pulse rifles) be an insta win item

**Comment by Mqiib** [comment]
Batons don't need to be buffed. They also don't need to be nerfed. As they are they're a perfectly useful tool to take people down when combined with other security gear. You can't solo a group of people with a single baton like you used to, which is a GOOD CHANGE. if you buff it to be a one-shot it doesn't matter how many hoops you make people jump through to get the upgrade, they'll do it and then shit on any and every antag around the moment they can get within melee range.

Lets do some math. If someone runs at you dual-wielding energy swords, they need to hit you four times to take you down. With a 75% block chance from DUAL WIELDING LIGHTSABERS they will now have a whopping 42% chance (0.75^3, three blocks in a row) to win that fight even if they get the first hit in, all else held equal. That is, a trained agent dual wielding fucking lightsabers and swinging them wildly at some brainless security officer with an electric stick has higher odds of LOSING the fight than winning. This chance drops to 31% (0.75^4, four blocks in a row) if the officer gets the drop on you, or 0% if you don't have your swords already out and ready when the officer runs in to hit you with their funny win stick. If you get whacked while still typing, you also immediately lose. It's dumb as hell and not even remotely resembling fun and interactive gameplay.

**Comment by Marmio64** [comment]
honestly would like to see the mindshield 'firing' pin for stunbaton, might slightly discourage antags from always just finding a loot pinata to kill instead of actually using their own gear. Would like to see this atomized so we could try the firing pin without the 1 hit stun upgrade.

**Comment by maxion12345** [comment]
half want this merged so I can go on a validhunt streak with them

anyways hopek has a point, what's the issues with our current batong that would necessitate this change? I've had no issues clicking people twice to stun (or once if using instant win cylinders) so maybe just uhhhhh git gud?

**Comment by ghost** [comment]
> Batons don't need to be buffed. They also don't need to be nerfed. As they are they're a perfectly useful tool to take people down when combined with other security gear. You can't solo a group of people with a single baton like you used to, which is a GOOD CHANGE. if you buff it to be a one-shot it doesn't matter how many hoops you make people jump through to get the upgrade, they'll do it and then shit on any and every antag around the moment they can get within melee range.

It actually needs to be nerfed. Yes, at this moment you can't run with them to a horde of people and win them. But you can run into a single guy and win him with two hits. But most common antags such as traitors, heretics, bloodbrothersare sole guys, so they will be get hardly fucked in most of cases.

**Comment by Hopekz** [comment]
> Yes, at this moment you can't run with them to a horde of people and win them. But you can run into a single guy and win him with two hits. But most common antags such as traitors, heretics, bloodbrothersare sole guys, so they will be get hardly fucked in most of cases.

That's the point. The antags you listed are supposed to fear sec vs having a balanced fight with them.

Otherwise what's the point of having sec in the game if they're only temporary roadblocks until you get good enough to 1v1 them every time. Sec is designed to win these 1v1 fights to encourage antags not fighting sec 1v1

If you could win 1v1 in those situations it wouldn't even matter if sec existed in the area you'd just do your objective in front of a sec officer then kill the sec officer in a 1v1, effectively making them pointless in the first place.

**Comment by ghost** [comment]
> That's the point. The antags you listed are supposed to fear sec vs having a balanced fight with them.
> 
> Otherwise what's the point of having sec in the game if they're only temporary roadblocks until you get good enough to 1v1 them every time. Sec is designed to win these 1v1 fights to encourage antags not fighting sec 1v1
> 
> If you could win 1v1 in those situations it wouldn't even matter if sec existed in the area you'd just do your objective in front of a sec officer then kill the sec officer in a 1v1, effectively making them pointless in the first place.

As an antagonist you are almost 1 vs everyone on the station, and sec mostly isn't getting in 1 vs 1 fights, and just swarms you with a group of funny armored guys that two-shot you with their battons. Because of this, making security that is supported by almost all the station, is containing many people that work in groups win 1 vs 1 fights is a questionable idea.

**Comment by stwordcitizen** [comment]
> > overpowered
> 
> How so?
stamina batons were fine, only people who were complaining about them were oldheads who needed their one hit uncounterable crutch baton. When this gets merged we'll be going right back to pre 2021 yogstation where the meta is to just steal a baton from the sec locker room and kill everyone with a one hit baton.

**Comment by JamieD1** [comment]
Some of these takes are awful, some actually have merit.

**Comment by Hopekz** [comment]
> Some of these takes are awful, some actually have merit.

You don't even have a take.

Your take is "I spent weeks discussing this with __"

What is the problem that caused this PR to exist?
How does this PR fix whatever problem it is trying to address?

File: code/game/objects/items/stunbaton.dm [hunk 17]
```diff
@@ -110,20 +129,59 @@
...
+			if(pin.type != /obj/item/firing_pin)
+				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
+				return
+		if(upgrade)
+			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
+		else
+			W.forceMove(src)
+			upgrade = W	
+	else if(W.tool_behaviour == TOOL_CROWBAR)
+		if(makeshift == TRUE)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
		if(makeshift)
```

File: code/game/objects/items/stunbaton.dm [hunk 18]
```diff
@@ -110,20 +129,59 @@
 			cell = W
 			to_chat(user, span_notice("You install a cell in [src]."))
 			update_icon()
-
+	else if(istype(W, /obj/item/firing_pin))
+		if(makeshift == TRUE)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
		if(makeshift)
```

File: code/game/objects/items/stunbaton.dm [hunk 19]
```diff
@@ -110,20 +129,59 @@
...
+			if(W.type != /obj/item/firing_pin)
+				to_chat("You are unable to add a non default firing pin whilst [src] has an upgrade. Remove the upgrade first with a crowbar.")
+				return
+		if(pin)
+			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar."))
+		else
+			W.forceMove(src)
+			pin = W
+	else if(istype(W, upgrade))
+		if(makeshift == TRUE)
```
Comment (Bibby0110) [inline code/game/objects/items/stunbaton.dm]:
```suggestion
		if(makeshift)
```

**Review by Bibby0110 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/stunbaton.dm [hunk 20]
```diff
@@ -268,6 +334,7 @@
 	hitcost = 2000
 	throw_hit_chance = 10
 	slot_flags = ITEM_SLOT_BACK
+	makeshift = TRUE
```
Comment (tattax) [inline code/game/objects/items/stunbaton.dm:368]:
```suggestion
```

File: code/game/objects/items/stunbaton.dm [hunk 21]
```diff
@@ -110,20 +129,59 @@
...
+				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
+				return
+		if(upgrade)
+			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
+		else
+			W.forceMove(src)
+			upgrade = W	
+	else if(W.tool_behaviour == TOOL_CROWBAR)
+		if(makeshift)
+			return
```
Comment (tattax) [inline code/game/objects/items/stunbaton.dm:178]:
```suggestion
		if(src.type != /obj/item/melee/baton)
			return
```

/////

```suggestion
		if(subtypesof(/obj/item/melee/baton))
			return
```

File: code/game/objects/items/stunbaton.dm [hunk 22]
```diff
@@ -110,20 +129,59 @@
...
+		if(pin)
+			if(pin.type != /obj/item/firing_pin)
+				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
+				return
+		if(upgrade)
+			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
+		else
+			W.forceMove(src)
+			upgrade = W	
+	else if(W.tool_behaviour == TOOL_CROWBAR)
```
Comment (tattax) [inline code/game/objects/items/stunbaton.dm:176]:
```suggestion
	else if(istype(W, /obj/item/firing_pin))
		if(src.type != /obj/item/melee/baton)
			return
		if(upgrade)
			if(W.type != /obj/item/firing_pin)
				to_chat("You are unable to add a non default firing pin whilst [src] has an upgrade. Remove the upgrade first with a crowbar.")
				return
		if(pin)
			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar."))
		else
			W.forceMove(src)
			pin = W
	else if(istype(W, upgrade))
		if(src.type != /obj/item/melee/baton)
			return
		if(pin)
			if(pin.type != /obj/item/firing_pin)
				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
				return
		if(upgrade)
			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
		else
			W.forceMove(src)
			upgrade = W	
	else if(W.tool_behaviour == TOOL_CROWBAR)
```

->switch(W.type)<-

```suggestion
	else if(istype(W, /obj/item/firing_pin))
		if(subtypesof(/obj/item/melee/baton))
			return
		if(upgrade)
			if(W.type != /obj/item/firing_pin)
				to_chat("You are unable to add a non default firing pin whilst [src] has an upgrade. Remove the upgrade first with a crowbar.")
				return
		if(pin)
			to_chat(user, span_notice("[src] already has a firing pin. You can remove it with crowbar."))
		else
			W.forceMove(src)
			pin = W
	else if(istype(W, upgrade))
		if(subtypesof(/obj/item/melee/baton))
			return
		if(pin)
			if(pin.type != /obj/item/firing_pin)
				to_chat("You are unable to upgrade the baton whilst it has a non default firing pin.")
				return
		if(upgrade)
			to_chat(user, span_notice("[src] already has an upgrade installed. You can remove it with crowbar"))
		else
			W.forceMove(src)
			upgrade = W	
	else if(W.tool_behaviour == TOOL_CROWBAR)
```

File: code/game/objects/items/stunbaton.dm [hunk 23]
```diff
@@ -34,6 +34,9 @@
 	var/preload_cell_type
 	///used for passive discharge
 	var/cell_last_used = 0
+	var/makeshift = FALSE
```
Comment (tattax) [inline code/game/objects/items/stunbaton.dm:38]:
```suggestion
```

**Comment by JamieD1** [comment]
https://forums.yogstation.net/threads/vote-908-merge-https-github-com-yogstation13-yogstation-pull-14306-after-code-quality-is-done.25306/unread

**Pull request merged**